src/java.base/share/classes/java/math/MutableBigInteger.java

```diff
@@ -1882,6 +1954,201 @@ MutableBigInteger sqrt() {
         }
     }
 
+    /**
+     * Calculate the integer square root {@code floor(sqrt(this))} and the remainder
+     * if needed, where {@code sqrt(.)} denotes the mathematical square root.
+     * The contents of {@code this} are <em>not</em> changed.
+     * The value of {@code this} is assumed to be non-negative.
+     *
+     * @return the integer square root of {@code this} and the remainder if needed
+     * @throws ArithmeticException if the value returned by {@code bitLength()}
+     * overflows the range of {@code int}.
+     */
+    MutableBigInteger[] sqrtRem(boolean needRemainder) {
+        // Special cases.
+        if (this.intLen <= 2) {
+            final long x = this.toLong(); // unsigned
+
+            /* For every long value s in [0, 2^32) such that x == s * s,
+             * it is true that s == Math.round(Math.sqrt(x >= 0 ? x : x + 0x1p64)),
+             * and if x == 2^64 - 1, then Math.round(Math.sqrt(x >= 0 ? x : x + 0x1p64)) == 2^32.
+             * This means that the value returned by Math.round(Math.sqrt())
```

Comment:
```suggestion
             * Since both `Math.round()` and `Math.sqrt()` are (weakly) increasing,
             * this means that the value returned by `Math.round(Math.sqrt())`
```